docs/Tutorial_MERLIN.ipynb

```diff
@@ -0,0 +1,407 @@
+{
```

Comment:
<p>Is there a reason why we are storing the subsequence length <code>m</code> inside of <code>D</code> ? It looks like it doesn't change and would be a constant. In fact, I would considering using two separate 1D arrays (<code>idx</code> and <code>distance</code>), do the computation, and combine the results into <code>D</code> at the end before returning. This would make the code a lot easier to read as the variable names are obvious. Eventually, I can help rename things to be more consistent with the variable names in the code base (i.e., we typically use <code>D</code> for a distance profile and so we'd likely just use <code>discords</code> to represent the set of discords, etc).</p><p> </p><p>Why are we converting <code>C</code> into a <code>list</code> ? Is this necessary?</p><p> </p><p>Instead of <code>D[j_ind,-1]&nbsp;</code> please be explicit and use <code>D[j_ind, 3]</code> .</p><p> </p><p>I noticed that <code>is_discord</code> isn't actually being used anywhere. You are setting it to <code>True</code>/<code>False</code> and that's it.</p><p> </p><p>The use of <code>C_indices_tot = np.arange(len(C_lst))</code> and <code>C_indices = C_indices_tot[:]</code> seem complicated to me. I wonder if there is a way to simplify all of this referencing. The Table 2 in the paper does not seem to look as complicated</p><p> </p><p>Why return a set of indices to remove? Why not just remove them before and only return the final set of <code>discords</code>  ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/418/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Reply:
> Is there a reason why we are storing the subsequence length `m` inside of `D` ?

I didn't understand this part. By subsequence, you mean the its (starting) index stored in the first column? If so, then the index itself is not going to change as I will keep them all and remove the unwanted ones later. However, its NN and its distance will be changed. So, if you are talking about `[j_ind,-1] = min(D[j_ind,-1], dist)`, then I think it is necessary as we try to update the NN distance of the discord located at index `j_ind`. 



> Why are we converting `C` into a `list` ? Is this necessary?
> 
> Instead of `D[j_ind,-1] ` please be explicit and use `D[j_ind, 3]` .
> 
> I noticed that `is_discord` isn't actually being used anywhere. You are setting it to `True`/`False` and that's it.

Thanks. I  will take care of it. In fact, that was when I tried to understand the difference between the proposed algorithm in the paper and the one provided in MATLAB and what `dj` is. So, I wrote as I tried to figure things out and didn't pay much attention to small details. So, I called it [WIP]. I will refine the code and will consider your comments in the next commit for sure.







> The use of `C_indices_tot = np.arange(len(C_lst))` and `C_indices = C_indices_tot[:]` seem complicated to me. I wonder if there is a way to simplify all of this referencing. The Table 2 in the paper does not seem to look as complicated
> 
> Why return a set of indices to remove? Why not just remove them before and only return the final set of `discords` ?

I totally agree. I need to go through it. I remember I got an error about the change in the size of `set` object  in the for loop. So, I tried to resolve it. I need to go back to that `set` approach.

Reply:
<p>Sounds good and no rush! I just wanted to provide feedback while it was still fresh in my head. I can tell you that I don't do well jumping back-and-forth from PR to PR. The context switching hurts my brain :)</p>

Reply:
> The Table 2 in the paper does not seem to look as complicated

![image](https://user-images.githubusercontent.com/38519522/123188560-754b9e80-d459-11eb-8eb8-9c81f951d937.png)

I tried to go through the algorithm one more time. I believe the Table 2 algorithm is not complete. Let's take a look at `dj`: It is initially set to `inf`. And, it is being updated through the outer loop . It is not shown in the TABLE 2; but, that is what authors did in their MATLAB code. They store it and updated it. (please see below)
```
% Algorithm 1: Discord Refinement Phase
cand_nndists2 = containers.Map(C.keys, inf(length(C), 1), 'UniformValues', true);
cand_nn_pos = containers.Map(C.keys, NaN(length(C), 1), 'UniformValues', true);
r2 = r^2;

% I diverge from the pseudocode in the paper here. First, I have to use 
% squared distances for correctness reasons in order to avoid taking a 
% square root at every step where we check against best so far. Second,
% I had to change the way elements are removed from the candidate set, or
% rather I had to change the loop tracking variable. If elements are 
% removed from an array while iterating over that same array, it will 
% result in skipped elements in almost any environment. Instead I'm 
% iterating over a copy of its hash keys, which remains constant over 
% the inner 2 loops. Elements are still progressively removed from the 
% candidate set.

for i = 1 : subseqcount
    if isempty(C)
        break;
    end
    keys = C.keys;
    Si = (TS(i : i + L - 1) - mu(i)) ./ sig(i);
    for j = 1 : length(keys)
        if abs(i - keys{j}) < min_separation
            continue;
        end
        dist2 = 0;
        candidx = keys{j};
        Candj = C(candidx);
        nndist2 = cand_nndists2(candidx); 
        for k = 1 : L
            dist2 = dist2 + (Si(k) - Candj(k))^2;
            if dist2 >= nndist2
                break;
            end
        end
        if dist2 < r2
            C.remove(candidx);
            cand_nndists2.remove(candidx);
            cand_nn_pos.remove(candidx);
            continue;
        elseif  dist2 < cand_nndists2(candidx)  
            cand_nndists2(candidx) = dist2;
            cand_nn_pos(candidx) = i;
        end
    end    
end
```

I can use dictionary to store `dj` (and also delete the ones (i.e. keys) that should  be removed throughout the for loop). I used numpy array instead for now. If you thin dictionary or any other structure is better to store/update the discords, pleases let me know.

========================================================================

Now, there is one thing left that should be noticed for now! I took a look at the documents on their webpage and realized: 

![image](https://user-images.githubusercontent.com/38519522/123189518-2e5ea880-d45b-11eb-8bc9-da73e23ec9ea.png)

There are a few differences:

(1) It says the NN is at index 470. But I didn't get that in the result of my implementation

(2)  [PROBABLY NOT IMPORTANT] The length of time series in the image above is 3000, similar to what provided in the paper. However, the length of the time series provided as a data set on the webpage of the paper is 2000. 

(3) It says the actual discord value is 10.27 (above which the algorithm returns no discord). However, my implementation shows that the actual discord value is 10.43.

===================================================================

I already revised the first two functions according to your suggestion. I will implement MERLIN and then push the commit.

Reply:
What value(s) do you get when you run the Matlab code?

Reply:
> What value(s) do you get when you run the Matlab code?

Will let you know as I haven't tried it yet.

Reply:
@udf2457 I was wondering if you had any thoughts on this?

Reply:
@seanlaw @udf2457 @alvii147 
So, I checked out the code again. As @alvii147 mentioned before, the Introduction Link contains a .zip file in which we can find the latest version of DRAG implemented by the author: `discord_discovery_gemm3.m`

I run the MATLAB code and compare it with mine using the same input:

The discord location and its distance to NN are the same. HOWEVER, the index of the NN subsequence are not the same.

Reply:
> The discord location and its distance to NN are the same. HOWEVER, the index of the NN subsequence are not the same.

Does that imply that there are two subsequences that have the same distance to the discord? Or do you think that they reported the wrong nearest neighbor index? Can you compute the "other" nearest neighbor distance by hand and compare?

Reply:
> > The discord location and its distance to NN are the same. HOWEVER, the index of the NN subsequence are not the same.
> 
> Does that imply that there are two subsequences that have the same distance to the discord? Or do you think that they reported the wrong nearest neighbor index? Can you compute the "other" nearest neighbor distance by hand and compare?

I just deleted my previous code. I am going to test the output of my implementation as I think it doesn't return the corrrect NN distance. (I checked the distance between subseq and its NN (provided in my output) and their distance is 0! I am now trying to debug it.

Reply:
Also, don't forget that you can find the top discord using the more expensive `stump` (i.e., compute the matrix profile and take the max). So you can at least use it to verify your results

Reply:
UPDATE:

EVERYTHING looks good! Thanks for your support. Now, I am going to do the MERLIN main algorithm,

Reply:
@seanlaw Just to say sorry for the delay in reply, yesterday was a bit nuts at $work.  @Ninimama good to hear you found a solution.